picard/ui/itemviews/custom_columns/script_provider.py

```diff
@@ -0,0 +1,143 @@
+# -*- coding: utf-8 -*-
+#
+# Picard, the next-generation MusicBrainz tagger
+#
+# Copyright (C) 2025 The MusicBrainz Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+"""Script-based provider with caching and performance thresholds."""
+
+from __future__ import annotations
+
+from collections import deque
+from collections.abc import Callable
+import re
+from time import perf_counter
+from weakref import WeakKeyDictionary
+
+from picard import log
+from picard.item import Item
+from picard.script import ScriptParser
+
+from picard.ui.itemviews.custom_columns.context import ContextStrategyManager
+from picard.ui.itemviews.custom_columns.resolve import ValueResolverChain
+
+
+class ChainedValueProvider:
+    """Provide script-evaluated values with caching and performance limits.
+
+    Caching strategy
+    ----------------
+    - Primary cache: WeakKeyDictionary keyed by the item when it supports
+      weak references. This avoids retaining objects strongly.
+    - Fallback id-cache: A bounded FIFO cache keyed by ``id(obj)`` for
+      objects that cannot be weakly referenced. Evictions use O(1)
+      ``deque.popleft``.
+    """
+
+    def __init__(
+        self,
+        script: str,
+        max_runtime_ms: int = 25,
+        cache_size: int = 1024,
```

Comment:
Maybe define those as `None` and use 2 class variables to store a default value:


```python
class ...
    DEFAULT_MAX_RUNTIME_MS = 25

    def ...
        max_runtime_ms: int | None = None,
        ...

        if max_runtime_ms is None:
            max_runtime_ms = self.DEFAULT_MAX_RUNTIME_MS
```

This way defaults can be changed more easily if needed (by subclassing or patching).